src/Simplex/Chat/Store/Messages.hs

```diff
@@ -1269,23 +1274,63 @@ getGroupChatScopeInfo_ db vr user GroupInfo {membership} = mapM $ \case
-  ts <- getCurrentTime
-  cis <- mapM (safeGetGroupItem db user g gcsi ts) ciIds
-  pure $ Chat (GroupChat g gcsi) (reverse cis) stats
+getGroupChatScopeInfoForItem :: DB.Connection -> VersionRangeChat -> User -> GroupInfo -> ChatItemId -> ExceptT StoreError IO (Maybe GroupChatScopeInfo)
+getGroupChatScopeInfoForItem db vr user g itemId = do
+  scope <- getGroupChatScopeForItem_ db itemId
+  getGroupChatScopeInfo' db vr user g scope
+
+getGroupChatScopeForItem_ :: DB.Connection -> ChatItemId -> ExceptT StoreError IO (Maybe GroupChatScope)
+getGroupChatScopeForItem_ db itemId =
+  ExceptT . firstRow toScope (SEChatItemNotFound itemId) $
+    DB.query
+      db
+      [sql|
+        SELECT group_scope_tag, group_scope_group_member_id
+        FROM chat_items
+        WHERE chat_item_id = ?
+      |]
+      (Only itemId)
+  where
+    toScope (scopeTag, scopeMemberId) =
+      case (scopeTag, scopeMemberId) of
+        (Just GCSTMemberSupport_, Just gmId) -> Just $ GCSMemberSupport gmId
+        (Just GCSTMemberSupport_, Nothing) -> Just $ GCSMemberSupport Nothing
+        (Nothing, Nothing) -> Nothing
+        (Nothing, Just _) -> Nothing -- shouldn't happen
+
+getGroupChatLast_ :: DB.Connection -> User -> GroupInfo -> Maybe GroupChatScopeInfo -> Maybe MsgContentTag -> Int -> String -> ChatStats -> ExceptT StoreError IO (Chat 'CTGroup)
+getGroupChatLast_ db user g scopeInfo_ contentFilter count search stats = do
+  ciIds <- getGroupChatItemIDs db user g scopeInfo_ contentFilter GRLast count search
+  ts <- liftIO getCurrentTime
+  cis <- mapM (liftIO . safeGetGroupItem db user g ts) ciIds
+  pure $ Chat (GroupChat g scopeInfo_) (reverse cis) stats
 
 data GroupItemIDsRange = GRLast | GRAfter UTCTime ChatItemId | GRBefore UTCTime ChatItemId
 
-getGroupChatItemIDs :: DB.Connection -> User -> GroupInfo -> Maybe GroupChatFilter -> GroupItemIDsRange -> Int -> String -> IO [ChatItemId]
-getGroupChatItemIDs db User {userId} GroupInfo {groupId, membership} groupChatFilter range count search = case groupChatFilter of
-  Nothing -> idsQuery (baseCond <> " AND group_scope_group_member_id IS NULL ") (userId, groupId)
-  Just (GCFChatScope Nothing) -> idsQuery (baseCond <> " AND group_scope_group_member_id IS NULL ") (userId, groupId)
-  Just (GCFChatScope (Just (GCSMemberSupport gmId_))) ->
-    let gmId = fromMaybe (groupMemberId' membership) gmId_
-     in idsQuery (baseCond <> " AND group_scope_group_member_id = ? ") (userId, groupId, gmId)
-  Just (GCFMsgContentTag mcTag) -> idsQuery (baseCond <> " AND msg_content_tag = ? ") (userId, groupId, mcTag)
+getGroupChatItemIDs :: DB.Connection -> User -> GroupInfo -> Maybe GroupChatScopeInfo -> Maybe MsgContentTag -> GroupItemIDsRange -> Int -> String -> ExceptT StoreError IO [ChatItemId]
+getGroupChatItemIDs db User {userId} GroupInfo {groupId} scopeInfo_ contentFilter range count search = case (scopeInfo_, contentFilter) of
+  (Nothing, Nothing) ->
```

Comment:
Looks like it should be a layer in this query constructor, the condition on scope is independent of the condition on content (and then you don't need to fail).